Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs

```diff
@@ -0,0 +1,131 @@
+    public class UnityDevice : HapticDevice
+    {
+#if UNITY_2018_3_OR_NEWER
+        #region Protected Variables
+        // Indicates which hand this object corresponds to
+        protected XRNode nodeType;
+        
+        // Holds a reference to the Unity XR device for haptic feedback
+        protected InputDevice hapticDevice;
+
+        // By default, this is set to zero
+        protected uint hapticChannel = 0;
+        #endregion
+
+        #region Haptic Device Behavior
+        protected override void CancelHaptics()
+        {
+#if UNITY_2019_1_OR_NEWER
+            if(hapticDevice.isValid)
+#elif UNITY_2018_3_OR_NEWER
+            if(hapticDevice.IsValid)
+#endif
+            {
+                hapticDevice.StopHaptics();
+            }
+            else
+            {
+                Debug.LogWarningFormat("Unity XR Haptic Device {0} failed to cancel: device is invalid", nodeType.ToString());
+            }
+        }
+
+        protected override void StartHaptics(HumanBodyBones bodyPart, BodyCoordinateHit hitLocation, float intensity)
+        {
+#if UNITY_2019_1_OR_NEWER
+            if(hapticDevice.isValid)
+#elif UNITY_2018_3_OR_NEWER
+            if (hapticDevice.IsValid)
+#endif
+            {
+                HapticCapabilities hapticCapabilities;
+                if (hapticDevice.TryGetHapticCapabilities(out hapticCapabilities) && hapticCapabilities.supportsImpulse)
+                {
+                    bool successful = hapticDevice.SendHapticImpulse(hapticChannel, intensity);
+                    if (!successful)
+                    {
+                        Debug.LogWarningFormat("Unity XR Haptic Device {0} failed to play haptics");
+                    }
+                }
+            }
+            else
+            {
+                Debug.LogWarningFormat("Unity XR Haptic Device for the {0} failed to play haptics: device is invalid", nodeType.ToString());
+            }
+        }
+
+        /// <summary>
+        /// This override is used to determine the correct XRNode type for this device.
+        /// </summary>
+        public override void ApplyDefaultData(HapticSystemAttribute hapticSystemInfo)
+        {
+            base.ApplyDefaultData(hapticSystemInfo);
+
+            nodeType = (bool)hapticSystemInfo.AdditionalData[0] ? XRNode.RightHand : XRNode.LeftHand;
+        }
+
+        /// <summary>
+        /// This override is called when the VRTK SDK is loaded. This is used here to fetch the Input Device handle for the appropriate controller.
+        /// </summary>
+        protected override void Instance_LoadedSetupChanged(VRTK_SDKManager sender, VRTK_SDKManager.LoadedSetupChangeEventArgs e)
+        {
+            if (e.currentSetup.controllerSDKInfo.type == hapticSystemInfo.ConnectedSDKType)
+            {
+                hapticDevice = InputDevices.GetDeviceAtXRNode(nodeType);
+            }
+        }
+        #endregion
+#else
+        protected override void CancelHaptics()
```

Comment:
can you then remove the overrides or is there value in having them here?
If purposeful, provide a leading comment.

Reply:
Added comment to clarify, but the base methods are abstract so need an implementation